File: Sources/LegacyClient/Public/Sources/Native/Entity/EntityType.ts
export const enum MobType {
    BEE,
    SPIDER,
    HORNET,

    BEETLE,
    SANDSTORM,
    CACTUS,
    SCORPION,
    LADYBUG_SHINY,

    STARFISH,
    JELLYFISH,
    BUBBLE,
    SPONGE,
    SHELL,
    CRAB,
    LEECH,

    CENTIPEDE,
    CENTIPEDE_EVIL,
    CENTIPEDE_DESERT,

    MISSILE_PROJECTILE,
    WEB_PROJECTILE,
}

export const enum PetalType {
    // Dont forgot to update to last index every time adding the type
    BASIC = MobType.WEB_PROJECTILE + 1,
    FASTER,
    EGG_BEETLE,
    BUBBLE,
    YIN_YANG,
    MYSTERIOUS_STICK,
    SAND,
    LIGHTNING,
    CLAW,
    FANG,
    YGGDRASIL,
    WEB,
    STINGER,
    WING,
}

export const PETAL_TYPES = [
    PetalType.BASIC,
    PetalType.FASTER,
    PetalType.EGG_BEETLE,
    PetalType.BUBBLE,
    PetalType.YIN_YANG,
    PetalType.MYSTERIOUS_STICK,
    PetalType.SAND,
    PetalType.LIGHTNING,
    PetalType.CLAW,
    PetalType.FANG,
    PetalType.YGGDRASIL,
    PetalType.WEB,
    PetalType.STINGER,
    PetalType.WING,
] as const;

File: Sources/Client/private/src/Florr/Native/Entity/EntityType.zig
const std = @import("std");
const meta = std.meta;

pub const MobType = enum(u8) {
    bee,
    spider,
    hornet,

    beetle,
    sandstorm,
    cactus,
    scorpion,
    ladybug_shiny,

    starfish,
    jellyfish,
    bubble,
    sponge,
    shell,
    crab,
    leech,

    centipede,
    centipede_evil,
    centipede_desert,

    missile_projectile,
    web_projectile,
};

const mob_type_fields = meta.fields(MobType);

pub const mob_type_last_index: u8 = mob_type_fields[mob_type_fields.len - 1].value;

pub const PetalType = enum(u8) {
    basic = mob_type_last_index + 1,
    faster,
    egg_beetle,
    bubble,
    yin_yang,
    mysterious_stick,
    sand,
    lightning,
    claw,
    fang,
    yggdrasil,
    web,
    stinger,
    wing,
    magnet,
};

pub const EntityType = union(enum(u8)) {
    mob: MobType,
    petal: PetalType,

    /// Returns direct entity type within this type.
    pub inline fn get(self: EntityType) u8 {
        switch (self) {
            .mob => |m| return @intFromEnum(m),
            .petal => |p| return @intFromEnum(p),
        }
    }

    pub inline fn isMob(self: EntityType) bool {
        return switch (self) {
            inline .mob => true,
            inline .petal => false,
        };
    }

    pub inline fn isPetal(self: EntityType) bool {
        return switch (self) {
            inline .mob => false,
            inline .petal => true,
        };
    }
};

pub const petal_types = [_]PetalType{
    .basic,
    .faster,
    .egg_beetle,
    .bubble,
    .yin_yang,
    .mysterious_stick,
    .sand,
    .lightning,
    .claw,
    .fang,
    .yggdrasil,
    .web,
    .stinger,
    .wing,
    .magnet,
};
